include/El/core/MemoryPool.hpp

```diff
@@ -0,0 +1,221 @@
+            }
+        }
+        alloc_to_bin_[mem] = bin;
+        return mem;
+    }
+    /** Release previously allocated memory. */
+    void Free(void* ptr)
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        auto iter = alloc_to_bin_.find(ptr);
+        if (iter == alloc_to_bin_.end())
+        {
+            RuntimeError("Tried to free unknown ptr");
+        }
+        else
+        {
+            size_t bin = iter->second;
+            alloc_to_bin_.erase(iter);
+            if (bin == INVALID_BIN)
+            {
+                do_free(ptr);
+            }
+            else
+            {
+                // Cache the pointer for reuse.
+                free_data_[bin].push_back(ptr);
+            }
+        }
+    }
+    /** Release all unused memory. */
+    void FreeAllUnused()
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (size_t bin = 0; bin < bin_sizes_.size(); ++bin)
+            for (auto&& ptr : free_data_[bin])
+                do_free(ptr);
+    }
+
+private:
+
+    /** Index of an invalid bin. */
+    static constexpr size_t INVALID_BIN = (size_t) -1;
+
+    /** Serialize access from multiple threads. */
+    std::mutex mutex_;
+
+    /** Size in bytes of each bin. */
+    std::vector<size_t> bin_sizes_;
+    /** Data available to allocate.
+     *  Each entry is a bin, and each bin has a vector of pointers to free
+     *  memory of that size.
+     */
+    std::vector<std::vector<void*>> free_data_;
+    /** Map used pointers to the associated bin index. */
+    std::unordered_map<void*, size_t> alloc_to_bin_;
+
+    /** Allocate size bytes. */
+    inline void* do_allocation(size_t size);
+    /** Free ptr. */
+    inline void do_free(void* ptr);
+
+    /** Return the bin index for size. */
+    inline size_t get_bin(size_t size)
+    {
+        // Assuming we don't have that many bins, just do a linear search.
+        // Could optimize with binary search if need be.
+        for (size_t i = 0; i < bin_sizes_.size(); ++i)
+        {
+            if (bin_sizes_[i] >= size) return i;
+        }
+        return INVALID_BIN;
+    }
+
+};  // class MemoryPool
+
+#ifdef HYDROGEN_HAVE_CUDA
+template <>
+inline void* MemoryPool<true>::do_allocation(size_t bytes)
+{
+    void* ptr;
+    auto error = cudaMallocHost(&ptr, bytes);
+    if (error != cudaSuccess)
+    {
+        RuntimeError("Failed to allocate CUDA pinned memory with message: ",
+                     "\"", cudaGetErrorString(error), "\"");
+    }
+    return ptr;
+}
+
+template<>
+inline void MemoryPool<true>::do_free(void* ptr)
+{
+    auto error = cudaFreeHost(ptr);
+    if (error != cudaSuccess)
+    {
+        RuntimeError("Failed to free CUDA pinned memory with message: ",
+                     "\"", cudaGetErrorString(error), "\"");
+    }
+}
+#endif  // HYDROGEN_HAVE_CUDA
+
+template <>
+inline void* MemoryPool<false>::do_allocation(size_t bytes) {
+    void* ptr = std::malloc(bytes);
```

Comment:
Googling seems to indicate that `new char[bytes]` is slightly preferable to `malloc(bytes)` for performance and safety reasons.

Reply:
Technically they pull from different resources. `malloc` pulls from the heap and `new` pulls from the free-store. (In reality, they are often the same, but we don't know that and the C++ standard doesn't say anything to force that.) The latter may be implemented in terms of the former, but not vice versa. If heap memory is conceptually what is needed here, `malloc` should be used; otherwise `new`. Also, if doing `new`, you should grab `unsigned char`, not `char`, which may or may not be signed.